### PR TITLE
returning the error when doing a nack in the message

### PR DESCRIPTION
**Pull request by cpanato**
Fixes: #831 

other protocol return the error, doing similar thing here as well

### CODE REVIEW

**Comment by duglin** [comment]
Seems reasonable. Is there a test we can add for this?

**Comment by duglin** [comment]
rebase is needed

**Comment by cpanato** [comment]
done @duglin, thanks for the review, PTAL

**Comment by duglin** [comment]
Is it possible to test that the error makes its way back to the logger?

**Comment by cpanato** [comment]
> Is it possible to test that the error makes its way back to the logger?
did some manual test and saw that showing up in the logs.

but not sure how to do that here in the tests, do you have a tests that setup the pubsub? sorry for asking that, i am too new to this codebase

**Comment by duglin** [comment]
I looked around and I didn't see a good example of us testing that yet so perhaps I'm asking too much.
Unless @embano1 can think of an easy way to test it, this
/LGTM

**Review by duglin — DISMISSED** [review]

**Comment by embano1** [comment]
> Is it possible to test that the error makes its way back to the logger?

Are you asking for the caller here, i.e., integration tests?

**Comment by duglin** [comment]
> > Is it possible to test that the error makes its way back to the logger?
> 
> Are you asking for the caller here, i.e., integration tests?

Yea, I guess an integration test since the testcase added (while valid) is kind of silly/trivial since it's basically just testing one if-statement, and the real point of this PR was to ensure that the error bubbled up to the proper spot (like the logger). So, in essence not just test the if-statement but also the Finish() logic in conjunction with our logging system. But if we don't have an existing test that can be easily hooked-into I'm not sure it's fair to ask @cpanato to add one just for this trivial PR.

**Comment by cpanato** [comment]
maybe we can open an issue for that, I think that will require some more complex things to setup

**Comment by duglin** [comment]
> maybe we can open an issue for that, I think that will require some more complex things to setup

Done: https://github.com/cloudevents/sdk-go/issues/979

I'm still ok with this PR. @embano1 for the final review/merge.

File: protocol/pubsub/v2/message_test.go [hunk 1]
```diff
@@ -45,3 +46,39 @@ func TestReadStructured(t *testing.T) {
...
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			msg := NewMessage(tc.pm)
+			err := msg.Finish(tc.err)
+			if tc.wantErr {
+				if err != tc.err {
+					t.Errorf("Error mismatch. got: %v, want: %v", err, tc.err)
+				}
+			}
```
Comment (embano1) [inline protocol/pubsub/v2/message_test.go:81]: nit: logic is partially broken IMHO if `tc.wantErr` is `false` but`err != nil`, i.e., the second test case does not seem to have any effect here?

Usually you'd do something like

```suggestion
			if tc.wantErr && err != nil {
				if err != tc.err {
					t.Errorf("Error mismatch. got: %v, want: %v", err, tc.err)
				}
			} else { // assert `tc.wantErr` is `false` and `err == nil` }
```

Comment (cpanato) [reply protocol/pubsub/v2/message_test.go:81]: yep, makes sense

Comment (cpanato) [reply protocol/pubsub/v2/message_test.go:81]: done, thanks for the review

**Review by embano1 — APPROVED** [review]
LGTM

**Pull request merged**